### PR TITLE
Rename Identifier -> Entity

### PR DESCRIPTION
**Pull request by WilliamDee**
resolves https://github.com/dbt-labs/dbt-semantic-interfaces/issues/9

### Description
<p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; color: rgb(230, 237, 243); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">In the new world of dbt-core x MetricFlow<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: var(--color-neutral-muted); border-radius: 6px;">Identifiers</code><span> </span>are becoming<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: var(--color-neutral-muted); border-radius: 6px;">Entities</code>. Additionally some of the properties of the object are changing. The resulting object should have the following properties</p>

Property Name | Type | Description
-- | -- | --
name | str | Name of the entity
type | enum | Type of the entity
description | str | Description of the entity
role | str | Role of the entity
entities | List[str] | List of composite sub-entities
expr | str | Expression of the entity

<p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px !important; color: rgb(230, 237, 243); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The above properties are were pulled from<span> </span><a class="issue-link js-issue-link" data-error-text="Failed to load title" data-id="5126938" data-permission-text="Title is private" data-url="https://github.com/dbt-labs/dbt-core/discussions/7456" data-hovercard-type="discussion" data-hovercard-url="/dbt-labs/dbt-core/discussions/7456/hovercard" href="https://github.com/dbt-labs/dbt-core/discussions/7456" style="box-sizing: border-box; background-color: transparent; color: var(--color-accent-fg); text-decoration: none; white-space: normal;">dbt-labs/dbt-core#7456</a>

<!--- 
  Before requesting review, please make sure you have:
  1. read [the contributing guide](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md),
  2. signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
  3. run `changie new` to [create a changelog entry](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: metricflow/dataset/convert_data_source.py [hunk 1]
```diff
@@ -294,16 +294,16 @@ def _convert_dimensions(
             select_columns=select_columns,
         )
 
-    def _create_identifier_instances(
+    def _create_entity_instances(
```
Comment (tlento) [inline metricflow/dataset/convert_data_source.py:297]: I am up to here.

**Review by tlento — COMMENTED** [review]
Working through it, need to make dinner. Renaming things is so terrible, sorry.

File: dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+class Entity(HashableBaseModel, ModelWithMetadataParsing):
+    """Describes a entity"""
+
+    name: str
+    description: Optional[str]
+    type: EntityType
+    role: Optional[str]
+    entities: List[CompositeSubEntity] = []
+    expr: Optional[str] = None
+
```
[the remark this answers is not in the document]
Comment (QMalcolm) [reply dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py]: Update: We've decided to keep it for now. We might not use it right away, and we might refine it slightly before we do. The metadata as a concept can be incredibly useful for user experience.

File: dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+class Entity(HashableBaseModel, ModelWithMetadataParsing):
+    """Describes a entity"""
+
+    name: str
+    description: Optional[str]
+    type: EntityType
+    role: Optional[str]
+    entities: List[CompositeSubEntity] = []
+    expr: Optional[str] = None
+    metadata: Optional[Metadata] = None
```
Comment (WilliamDee) [inline dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py:48]: We had it as `Optional[Metadata]` without a default value. I added the default of `None` to allow not having to explicitly set it to `None` if we don't pass this property in.

File: dbt_semantic_interfaces/pyproject.toml [hunk 4]
```diff
@@ -24,3 +24,23 @@ more-itertools = "8.10.0"
 [build-system]
 requires = ["poetry-core>=1.0.0"]
 build-backend = "poetry.core.masonry.api"
+
+[tool.black]
```
Comment (WilliamDee) [inline dbt_semantic_interfaces/pyproject.toml:28]: @QMalcolm added this cause there was some CI issues with different black configurations between MF and this repo.

File: metricflow/test/model/parsing/test_data_source_parsing.py [hunk 5]
```diff
@@ -157,71 +156,9 @@ def test_data_source_identifier_parsing() -> None:
     assert identifier.name == "example_identifier"
     assert identifier.type is EntityType.PRIMARY
     assert identifier.role == "test_role"
-    assert identifier.entity == "other_identifier"
     assert identifier.expr == "example_id"
 
 
-def test_data_source_identifier_metadata_parsing() -> None:
```
Comment (QMalcolm) [inline metricflow/test/model/parsing/test_data_source_parsing.py:164]: since we're bringing back metadata, this test should come back as well.

**Review by QMalcolm — COMMENTED** [review]
Review of first 10 commits. Will review the rest after lunch.

File: dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (QMalcolm) [reply dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py]: Update: We've decided to keep it for now. We might not use it right away, and we might refine it slightly before we do. The metadata as a concept can be incredibly useful for user experience.

File: dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    name: str
+    description: Optional[str]
+    type: EntityType
+    role: Optional[str]
+    entities: List[CompositeSubEntity] = []
+    expr: Optional[str] = None
+
+    @property
+    def is_primary_time(self) -> bool:  # noqa: D
```
[the remark this answers is not in the document]
Comment (WilliamDee) [reply dbt_semantic_interfaces/dbt_semantic_interfaces/objects/elements/entity.py:51]: From looking at the usages of this, I believe the answer to that is no. Will remove. @tlento to confirm

**Review by tlento — COMMENTED** [review]
Whew. Massive!

Test assertion change is odd, what's up there?

File: metricflow/test/model/test_data_source_container.py [hunk 7]
```diff
@@ -197,4 +197,4 @@ def test_local_linked_elements_for_metric(metric_semantics: MetricSemantics) ->
 def test_get_data_sources_for_entity(data_source_semantics: DataSourceSemantics) -> None:  # noqa: D
     entity_reference = EntityReference(element_name="user")
     linked_data_sources = data_source_semantics.get_data_sources_for_entity(entity_reference=entity_reference)
-    assert len(linked_data_sources) == 9
+    assert len(linked_data_sources) == 8
```
[the remark this answers is not in the document]
Comment (WilliamDee) [reply metricflow/test/model/test_data_source_container.py:200]: So basically, when I removed the entity property on the `Identifier` class, I removed all the `entity` fields on the test models. One of them was this https://github.com/dbt-labs/metricflow/pull/490/commits/200445ca5c829a5c0eedfa043d149b79fe5729fe#diff-e939351cfbeb60d978e42432cbfb25ebba45ff60f4ca48457712dad0eefd0ce3 and from `get_data_sources_for_entity("user")`, that `bookings_source` data source is no longer returned from that function call.

File: metricflow/test/model/validations/test_entities.py [hunk 8]
```diff
@@ -231,20 +231,21 @@ def test_mismatched_identifier(simple_model__with_primary_transforms: UserConfig
...
-        identifiers=[CompositeSubEntity(ref="sub_identifier2")],
+        entities=[CompositeSubEntity(ref="sub_identifier2")],
     )
     listings_latest.identifiers = tuple(more_itertools.flatten([listings_latest.identifiers, [identifier_listings]]))
 
     model_issues = ModelValidator([EntityConsistencyRule()]).validate_model(model)
 
-    expected_error_message_fragment = "does not have consistent sub-identifiers"
+    expected_error_message_fragment = "does not have consistent sub-entities"
+    print(model_issues.all_issues)
```
[the remark this answers is not in the document]
Comment (WilliamDee) [reply metricflow/test/model/validations/test_entities.py]: woops, good catch!

File: metricflow/dataflow/builder/node_evaluator.py [hunk 9]
```diff
@@ -144,29 +144,27 @@ def _find_joinable_candidate_nodes_that_can_satisfy_linkable_specs(
...
-                    if instance.spec == identifier_spec_in_right_node:
+                    if instance.spec == entity_spec_in_right_node:
                         identifier_instance_in_right_node = instance
                         break
 
                 if identifier_instance_in_right_node is None:
-                    raise RuntimeError(
-                        f"Could not find identifier instance with name ({identifier_spec_in_right_node})"
-                    )
+                    raise RuntimeError(f"Could not find identifier instance with name ({entity_spec_in_right_node})")
```
Comment (QMalcolm) [inline metricflow/dataflow/builder/node_evaluator.py]: `f"Could not find identifier ...` -> `"Could not find entity ...`

**Review by QMalcolm — COMMENTED** [review]
Finished my first pass! 😂

File: metricflow/dataflow/builder/node_evaluator.py (hunk 9 above)
Comment (QMalcolm) [reply metricflow/dataflow/builder/node_evaluator.py]: Looks like you caught this already, nice!

File: metricflow/test/model/test_data_source_container.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlento) [reply metricflow/test/model/test_data_source_container.py:200]: Ok cool, that makes sense, thanks!

**Review by tlento — APPROVED** [review]
LGTM, assuming all of the various snapshot tests pass across engines and @QMalcolm doesn't turn up any issues.

**Review by QMalcolm — APPROVED** [review]
I think we should dub this "The Great Renamen-ing". This looks good to me! (contingent that the remaining references to `identifers` / `identifer` are planned to go away in the renaming of `DataSource` to `SemanticModel`

**Pull request merged**